tests/sles4sap/netweaver_ascs.pm

```diff
@@ -0,0 +1,79 @@
+# SUSE's SLES4SAP openQA tests
+#
+# Copyright © 2017 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Checks NetWeaver's ASCS installation as performed by sles4sap/nw_ascs_install
+# Requires: sles4sap/nw_ascs_install
+# Maintainer: Alvaro Carvajal <acarvajal@suse.de>
+
+use base "x11test";
+use strict;
+use testapi;
+
+sub run {
+    my ($self) = @_;
+    my $pscmd = "ps auxw | grep ASCS | grep -vw grep";
+    $pscmd = "$pscmd | wc -l ; $pscmd";
+
+    x11_start_program('xterm');
+    assert_screen('xterm');
+
+    # The SAP Admin was set in sles4sap/nw_ascs_install
+    my $sapadmin = get_var('SAPADM');
+    die "netweaver_ascs: coulnd't determine the SAP Administrator's username"
+      unless ($sapadmin);
+
+    # Allow SAP Admin user to inform status via $testapi::serialdev
+    assert_script_sudo("chown $sapadmin /dev/$testapi::serialdev", 5);
+
+    type_string "su - $sapadmin\n";
+    type_string "$testapi::password\n" unless ($testapi::username eq 'root');
+
+    assert_script_run("sapcontrol -nr 00 -function GetVersionInfo", 20);
+    assert_screen('netweaver-version-info', 10);
+
+    type_string "clear\n";
+    assert_script_run("sapcontrol -nr 00 -function GetInstanceProperties | grep ^SAP", 20);
+    assert_screen('netweaver-instance-properties', 10);
+
+    type_string "clear\n";
+    assert_script_run("sapcontrol -nr 00 -function Stop", 20);
+    assert_screen('netweaver-stop-instance', 10);
+
+    type_string "clear\n";
+    assert_script_run("sapcontrol -nr 00 -function StopService", 20);
+    assert_screen('netweaver-stop-service', 10);
+
+    type_string "clear\n";
+    type_string "$pscmd\n";
+    check_screen('netweaver-service-started', 30);
```

Comment:
Why? Does it break something? I am not interested in stopping the test at that point if that 'ps' command fails or if it doesn't show what I expected. As you said, VMs are slow, and the 2 previous sapcontrol commands stopping the instance and stopping the service while successful (I'm checking both with assert_script_sudo() and assert_screen()) could still have processes running in the BG that could show (or not) on 'ps's output. But I do want to check the screen so the 'ps' command output is available to the reviewer. I guess I could change the type_string to script_run and remove the call to check_screen, but thought it looked better this way.

Reply:
if you not interested in check why you have it ? just delete this line at all. The point is that it doesn't make any sense to ask question if you don't care about an answer :)

Reply:
http://open.qa/api/testapi/#_check_screen also mentions "Us this for optional matches", means "check the return value". I'm just speaking from experience. I have seen the following symptoms:
* script_run's or check_screen's fail pointing to horrible, unexpected problems and no one noticed or gets very confused only looking at a later step which then fails but is not easily understandable because the error was ignored in the previous script_run/check_screen
* Other people will use the code as reference and think it is good to use that style to write tests
* In case of expected "false" results the red border around screenshots is still confusing reviewers thinking that this is a cause of a problem when actually it always failed and you expected it to fail hence used "only" check_screen.

If you are just interested in seeing a screenshot there is also the test API command `save_screenshot`. In other cases when you don't want to be too strict then use assert_script_run or assert_screen with a very high timeout. About needles try to cover as few areas as possible so that your screen check is resilient and not confused by screen content that you don't care about.

Reply:
Yes, I am interested in the screen or the output being visible to the reviewer but I did try with `save_screenshot` and didn't do what I needed, so I went with `check_screen` instead. Thanks for the recommendation regarding script_run/check_screen. I think I will go with an assert_script_run then.